### PR TITLE
Adds https support with self-signed certs

### PR DESCRIPTION
**Pull request by tannerjfco**
## The Problem:
drud/ddev#232 OP - We want to support both HTTP and HTTPS for local dev without redirect.

## The Fix:
This (re-)introduces support for HTTPS to ddev router. jwilder/nginx-proxy, the project ddev router is based on, has provided HTTPS support for a while, but we removed most of this from our templating, as we weren't using it and were dealing with other complexities in the template at the time.

The new support is provided by introducing an HTTPS_EXPOSE env var, which is used to define the server blocks for HTTPS. This leverages upstreams created by defining HTTP_EXPOSE. The end result is customizable HTTPS port serving the desired HTTP backend. The other component to this functionality is generation of self-signed certs, which are generated as wildcards for any detected VIRTUAL_HOST hostnames.

Since the time that we originally split off from the upstream project, several HTTPS/SSL related improvements have made their way into nginx-proxy. Because of this, and the difficulties I hit re-implementing on our current template, I opted to start over from the current upstream template, and re-implement our customizations onto that. While I was at it, I made sure to remove template functions we aren't using, and attempted to better document our customizations. Hopefully this can allow for a bit less pain the next time we have to do something with the router template.

Not great things:
- upstream nginx-proxy hasn't pushed any new tagged releases since April, and a fair bit has gone into nginx-proxy since then. I've pushed a tagged image to our dockerhub to facilitate a stable upstream image for ddev-router. Hopefully they will push a versioned tag in the near future so we can lean on that instead.

## The Test:
This should be tested as part of the ddev PR TBD

### CODE REVIEW

**Comment by rfay** [comment]
This should include a self-signed wildcard certificate and key, configured on port 443 for all sites based on my understanding of what we're doing.

**Comment by rfay** [comment]
The other very significant advantage of a single wildcard cert is that the browser may only have to accept it once for all sites (I'm not certain of this and haven't tested it with browsers).

**Comment by tannerjfco** [comment]
The cert that is generated is a wildcard for *.ddev.local and gets used for any such subdomains. This achieves the same effect that a single checked-in wildcard cert would. I ended up going with a generation approach here for a few reasons:

- This allows greater flexibility for the domain to change as needed. Going ddev.local->ddev.site won't require touching the router again, nor should allowing users to define their own domain entirely, should that be a direction we end up going.
- I didn't see significant difference in generating the cert as part of docker build vs generating dynamically. Either way I need a non-interactive openssl command, and the supporting templating for dynamic generation is not all that complex.
- I was encountering unexpected removal of certs after building them in. I couldn't determine what would be removing them, but at this point I already had dynamic generation in mind and didn't want to chase down or work around that for a more rigid static asset.
- We won't be worrying about certs expiring this way. This isn't as huge a concern, as its unlikely we'll go over a year without touching ddev-router, but this way we always have fresh certs without a maintenance requirement on the image.

**Comment by tannerjfco** [comment]
> The other very significant advantage of a single wildcard cert is that the browser may only have to accept it once for all sites (I'm not certain of this and haven't tested it with browsers).

I don't believe this is the case. The security exception is for the domain itself, not the certificate.

**Comment by rfay** [comment]
However, the generated cert is not persistent, true? Meaning that people get annoyed every time they router has shut down? Or is it every time a site is shut down and restarted?

**Comment by rfay** [comment]
Actually, though, with a wildcard, you *may* be accepting validity for the entire wildcard domain.

**Comment by tannerjfco** [comment]
> However, the generated cert is not persistent, true? Meaning that people get annoyed every time they router has shut down? Or is it every time a site is shut down and restarted?

I'm not sure where user annoyance factors in here. The generated cert is as persistent as the nginx configuration - it exists as long as the router runs and there are site containers to serve. The generation of the cert should be as visible to users as the nginx config generation, in that it is not.

> Actually, though, with a wildcard, you may be accepting validity for the entire wildcard domain.

I tested this in Chrome and Firefox with 2 subdomains off ddev.local. For both browsers I had to provide separate exceptions. This indicates to me that the exception is for the specific domain you are attempting to access, not the certificate the site is trying to use.

**Comment by rfay** [comment]
I'm afraid I can't support the non-persistent generated cert approach. The issue called for a static wildcard certificate, and it's important to at least my workflow. I often work on many sites, shutting them down, or shutting down all sites, then expect to be able to come back and work on them without any change to my workflow (like accepting a new cert). 

Please do this with a static self-signed wildcard certificate. 

I'd be *way* in favor of the ability to mount legitimate certs from the host but that's probably down the line somewhere, but it's worth keeping in mind.

**Comment by tannerjfco** [comment]
I have a hard time getting on board with baking into the container an expiring asset tied to a specific domain. Doing so effectively puts an expiration date on the given tag for a container, at the very least a date at which the tag will start having new SSL-related issues.

I've updated this and the ddev PR to instead mount a volume for certs, allowing them to persist. The cert will then be checked for expiry within 48hrs, and will be regenerated if so. This should achieve the following:

- Router shutting down/restarting does not create new exemption requirement in the browser.
- Similarly, our updating of the router image won't create a new exemption requirement in the browser.
- Users shouldn't encounter a cert expiring.

**Comment by rfay** [comment]
I think that's  a step forward, but could you please make the *wildcard* cert expire at the *maximum* time, like 9 years or whatever it is (don't remember). There's no reason for it to even expire. I'm fine with checking it, but there's no reason I understand for it to expire in any normal time period. I like that you're doing generation though, it's a bit nicer than a static cert, as long as it doesn't add too much complexity. 

Is it possible then to make this be replaced with a legit cert if people have one (or if we find an easy way to provide one)?

**Comment by rfay** [comment]
BTW, if these are separate certs and you prefer it that way, you're convinced this is simple enough to be maintainable, and especially if the cert for a site can be swapped for a real one, I'm ok with separate certs. I just thought wildcard was so very much simpler.

**Comment by tannerjfco** [comment]
Thanks, I've updated the wildcard cert length to 9 years. I didn't realize certs could be made for longer than a couple years, likely because CAs don't tend to do more than 2 years.

This work should allow just fine for a user to replace our generated cert with one of their own, so long as their cert is not close to expiration. A user could replace the ddev.local.crt generated, or they could provide a specificsub.ddev.local.crt to use a custom cert for a specific site/subdomain

File: Dockerfile.in [hunk 1]
```diff
@@ -1,17 +1,32 @@
...
+RUN apt-get -qq update && \
+    apt-get -qq install --no-install-recommends --no-install-suggests -y \
+        ca-certificates wget vim less telnet curl procps && \
+    apt-get autoremove -y && \
+    apt-get clean -y && \
 	rm -rf /var/lib/apt/lists/*
 
+ADD https://github.com/jwilder/forego/releases/download/v0.16.1/forego /usr/local/bin/forego
+
+RUN wget https://github.com/jwilder/docker-gen/releases/download/$DOCKER_GEN_VERSION/docker-gen-linux-amd64-$DOCKER_GEN_VERSION.tar.gz && \
```
Comment (rfay) [inline Dockerfile.in]: Would love it if this could be quieted. It makes a lot of noise.

**Pull request merged**